[Pre-MAP Course Website](http://depts.washington.edu/premap/seminar/cohort-17-2021-seminar/) | [Pre-MAP GitHub](https://github.com/UWPreMAP/PreMAP2021) | [Google](https://www.google.com)

### Each time you access the PreMAP2021 directory make sure your files are up to date
1. Open up a terminal tab (New -> Terminal). Change directories into the PreMAP2021 directory, then do:
```bash
cd PreMAP2021
```
2. Update the directory to get any newly added files by running in the terminal:
```bash
git pull
```
3. Type in your terminal:
```bash
cd lessons
```
4. If you're on the AstroLab computer, type in your terminal:
```bash
jupyter notebook
```
This will open a webpage that has the lessons on them. You can select this lesson and then edit and run the cells to follow along with the lesson. Remember to change "Lastname" to your last name.

# Control flow/logic
### Control flow statements literally control the flow of your code!
There are a few basic types of control flow statements that we will learn. These are as follows: 
- **if:** do something _if_ some condition is `True`
- **for:** do something repeatedly _for_ some number of times, or on some number of items
- **while:** do something _while_ some condition is `True`

Let's look at some examples. **Notice that indentation (whitespace preceding code) matters in Python**! Blocks of text following a `for`, `while` or `if` statement must be indented to run properly. You can think of the indent as saying "do".

## Boolean review

Remember that a boolean is a special data type in python that has two possible values, `True` or `False`.

When we type a statement in python that results in a Boolean, we are essentially asking if that statement is `True` or `False`. For example

In [1]:
number = 5

print(number > 2)
print(number < 2)

True
False


This doesn't just hold for numbers, we can check if a variable is equal to a string as well. Remember that when we want to check if two things are equal, we use `==`.

In [2]:
chalkboard_text = "The Simpsons"

print(chalkboard_text == "Futurama")

False


This will come in handy today as we learn how to write code that only runs sometimes to handle multiple cases.

### `and` and `or`

You can make complex statements that result in Booleans by using `and` and `or`. These work as follows:

In [3]:
number = 5

In [4]:
print(number > 0 and number == 5)

True


In [5]:
print(number > 0 and number > 10)

False


In [6]:
print(number < 0 and number > 10)

False


In [7]:
print(number > 0 or number == 5)

True


In [8]:
print(number > 0 or number > 10)

True


In [9]:
print(number < 0 or number > 10)

False


## The `if` statement

When you need a program to do something more complicated than arithmetic, you'll sometimes need the code to do something _if_ some condition is met, otherwise it should do something else.

The syntax is as follows

```python
if STATEMENT:
    THING
```

so, python does the THING if the STATEMENT is true. We can also tell python what to do if STATEMENT is not true, with:

```python
if STATEMENT:
    THING
else:
    OTHERTHING
```

One basic example is checking a password. _If_ someone submits the right password, then tell them they're welcome, _else_ (otherwise) give them a warning to get away. Here's how we can do that in Python:


In [10]:
password = 'friend'
truePassword = 'friend'

if password == truePassword:
    # This indented code will only run if the above condition is True
    print('Enter.')
else:
    # This indented code will only run if the above condition is False
    print('You shall not pass!')

Enter.


The above example has only two possible outcomes - either the password is correct or incorrect. Sometimes, you'll want more possible outcomes, and for those occasions you can use the `elif` statement, meaning "else-if". For example, let's create a program that tells you how someone might react to a gift of skittles:

Run the code in the cell below and see what happens when you change the value of the `skittles_gifted` variable.

In [11]:
skittles_gifted = 3

if skittles_gifted < 5:
    print("I would love some more skittles if you'll share.")
elif skittles_gifted < 10:
    print("Maybe one more skittle please.")
elif skittles_gifted < 50: 
    print("Thank you, but this is too many skittles.")
else:
    print("Oh no! I'm buried in skittles now.")

I would love some more skittles if you'll share.


### Example 1

In the cell below, explain the sequence of events in the code starting from the first line, when `course_number = 601`, in comments: 

In [12]:
course_number = 601

if course_number < 200:
    print("This course is going to be fun.")
elif course_number < 300:
    print("This course is going to be hard.")
elif course_number < 400:
    print("You might want to wait for this course.")
else:
    print("Runaway!")

Runaway!


### Example 2

In the cell below, create a program that checks that the user paid the correct amount for a product. You should make two variables, `customer_paid` for how much the customer gave you, and `price` for how much the product costs. Then your program will do one of three things: 
1. If the customer pays exactly the price, print "thanks!"
2. If the customer pays more than the price, print "thanks!!!"
3. If the customer pays less than the price, print "try again"

In [13]:
customer_paid = 2# euros
price = 3 # euros

if customer_paid < price:
    print("try again")
elif customer_paid == price:
    print("thanks!")
elif customer_paid > price:
    print("thanks!!!")

try again


## The `while` loop

The `while` command is the first type of _loop_ we'll use. A loop is a snippet of code that gets run repeatedly. 

The syntax is as follows
```python
while STATEMENT:
    THING
```
This tells python to do the THING while STATEMENT is true.

Let's see an example below, where we use a `while` loop to add 1 to `number` for as many times as necessary until `number` is greater than or equal to 10:

Run the code in the cell below. What gets printed and why?

In [14]:
number = 0

while number < 10:
    print(number)
    number += 1

0
1
2
3
4
5
6
7
8
9


### Example 4

Using the example above as a template, create a variable `counter = 5000`, and use a `while` loop to subtract 10 from `counter` until `counter` is less than -5000. 

Print `counter` after the loop is complete, so you can see the result.

In [15]:
counter = 5000

while counter > -5000:
    counter = counter - 10

print(counter)

-5000


***

It's possible to accidentally create a loop that has no end. For example, if you wrote the above loop like this,
```python
counter = 5000

while counter < 6000:
    counter -= 10
```
the `counter` will always be less than 6000, so the condition `counter < 6000` will always be `True`, and the loop will never end. That's called an _infinite loop_. If you create an infinite loop, the program will keep running forever, unless you _kill_ it. You can kill a Python cell in an iPython Notebook by pressing the Stop button in the toolbar at the top of the page. 

### Example 5

In this example, we'll create an infinite loop that doesn't go on forever, because it terminates with a `break` statement. `break` forces python to end a loop.

Let's start a `counter` at 10, and iteratively subtract one from `counter`. We could make the `while` loop terminate when `counter == 1` by writing the loop like this
```python
counter = 10

while counter > 1:
    counter -=1
    
print(counter)
```
Or, we could `break` out of an infinite loop:
```python
counter = 10

# This loop is infinite on purpose:
while True:
    counter -= 1
    
    # When counter is equal to one, break out of the loop:
    if counter == 1:
        break

print(counter)
```

### Example 6

1. Copy and paste the loop with the `break` from above into the cell below, and check that it works.

2. Modify the loop so that it terminates when the `counter` is at negative one million (reminder: you can use `e` for scientific notation)

3. Now modify the loop so that it counts down from 10 to negative one million by twos.

In [16]:
counter = 10

# This loop is infinite on purpose:
while True:
    counter -= 1

    # When counter is equal to one, break out of the loop:
    if counter == 1:
        break

print(counter)

1


### The `for` loop

The `for` loop is trickier to learn than the other statements we've learned so far, but it's worth the trouble! 

A `for` loop is used to do something with every element of a list or array. The syntax is as follows

```python
for VARNAME in ITERABLE:
    DO THING WITH VARNAME
```
You can give VARNAME whatever name you want. When you do something with that variable name in the for loop, python will do that to every element of ITERABLE.

Let's say you have a grocery list of items you want to buy from the store. And let's say that the store publishes a list of prices for each item.

In [17]:
grocery_list = ['eggs', 'tofu', 'milk', 'chorizo', 'potatoes', 'tortillas', 'candy']

# List is in dollars for (in order): eggs, tofu, milk, bread, candy
prices_list = [1.99, 2.99, 0.99, 3.99, 1.99, 0.99, 5.99] 

print(grocery_list, prices_list)

['eggs', 'tofu', 'milk', 'chorizo', 'potatoes', 'tortillas', 'candy'] [1.99, 2.99, 0.99, 3.99, 1.99, 0.99, 5.99]


Note that the lists are printed sequentially. It would be a lot easier to read this like a table, with a column of items, and a column of their corresponding prices. We can do that with the code below. 

In [18]:
print(grocery_list[0], prices_list[0])
print(grocery_list[1], prices_list[1])
print(grocery_list[2], prices_list[2])
print(grocery_list[3], prices_list[3])
print(grocery_list[4], prices_list[4])
print(grocery_list[5], prices_list[5])
print(grocery_list[6], prices_list[6])

eggs 1.99
tofu 2.99
milk 0.99
chorizo 3.99
potatoes 1.99
tortillas 0.99
candy 5.99


But that took a lot of typing! And would take even more typing if we had a longer grocery list. We can do this really efficiently with a `for` loop. But we have to build up to it first!

Let's start by printing each name in the `grocery_list` list with a `for` loop: 

In [19]:
for grocery in grocery_list: # Think about this as saying "for each name in grocery_list: do"
    print(grocery)

eggs
tofu
milk
chorizo
potatoes
tortillas
candy


Let's break it down. `grocery_list` is the list that we're going to _iterate_ over, or _loop_ over. When you say `for grocery in grocery_list`, here's what happens: 
* the `for` says you're going to start a `for` loop
* `grocery` is a new variable that you're specifying, which the `grocery` loop is going to give values to. You can name this variable whatever you want
* `in` says that each `grocery` is going to come from an element of the `grocery_list` list
* `grocery_list` is the list that you're going to pull elements from

Here's what is going to happen when the `for` loop is run:
1. The `for` loop sets the variable `grocery` equal to the first item in `grocery_list` (which is `"eggs"`)
2. The indented code is executed
3. The `for` loop sets `grocery` equal to the second item in `grocery_list` (which is `"tofu"`)
4. The indented code is executed
5. The `for` loop sets `grocery` equal to the third item in `grocery_list` (which is `"milk"`)
6. Repeat until you've iterated over every element in `grocery_list`

### Example 7

In the cell below, use the `for` loop above as a template to create a `for` loop that prints the price of each grocery from the `prices_list`.

In [20]:
for price in prices_list:
    print(price)

1.99
2.99
0.99
3.99
1.99
0.99
5.99


## Making a table of groceries and their prices

So, we want to make a table that has two columns, the first column has the groceries and the second has their prices. You might think, "if a for loop iterates over one list, then how can we do something with two lists?"

This is where a new function called `range` comes in handy. Let's see what it does in the context of `for` loops.

In [21]:
for i in range(5):
    print(i)

0
1
2
3
4


The `for` loop is first setting `i=0` and printing `i`, then it sets `i=1` and prints `i`, and keeps going until the last element, 4.

The range function is producing a series of consecutive integers just like `np.arange` did. The argument `5` tells `range` to produce the numbers from zero to four.

Now, let's recall how indexing of lists works. If we want to see the first element of the `grocery_list` list we would write `grocery_list[0]`, or `grocery_list[1]` for the second element of the list. 

So, let's print every grocery in the `grocery_list` with a `for` loop and indexing.

In [22]:
number_of_groceries = len(grocery_list)
print(number_of_groceries)

7


In [23]:
for number in range(number_of_groceries):
    print(grocery_list[number])

eggs
tofu
milk
chorizo
potatoes
tortillas
candy


So, you might think "Hey! This is just the same as the code we wrote above where we said
```python
for grocery in grocery_list:
    print(grocery)
```
what's the point of doing it a different way?"

Well, in this case, you're not iterating over ``grocery_list``, you're iterating over the indices of ``grocery_list``, and these indices can be used to index into other lists.

Let's get back to our goal, printing each grocery and their corresponding price in two columns.

### Example 8

Use a single for loop, the ```range``` function, and indexing to print each grocery and their corresponding price in a two-column format.

**Heads up:** this is the toughest coding we've done so far, and it's going to be confusing. Nobody understands `for` loops the first time that they see them, so don't hesitate to ask for help. Remember to talk with your neighbors, and to raise your hand if you get stuck!

In [24]:
for number in range(number_of_groceries):
    print(grocery_list[number], prices_list[number])

eggs 1.99
tofu 2.99
milk 0.99
chorizo 3.99
potatoes 1.99
tortillas 0.99
candy 5.99


### Example 9

`for` loops are mostly useful for doing repeated calculations on items of a list. 

Let's assume that the store is having a sale, and all prices are cut in half!

In the cell below, make a `for` loop that prints how much you'll actually pay for each price in ``prices_list``.

In [25]:
for price in prices_list:
    print(price / 2)

0.995
1.495
0.495
1.995
0.995
0.495
2.995


### Making a list with for loops

So far, we've seen examples where we wanted to get elements out of a list, but we also use `for` loops to add elements to lists. 

Let's make a new list of prices that represents how much you'll actually pay for each grocery.

Without a `for` loop you would do that like this.

In [26]:
actual_prices = [] # Make an empty list that we will add our discount prices to

actual_prices.append(prices_list[0] / 2)
actual_prices.append(prices_list[1] / 2)
actual_prices.append(prices_list[2] / 2)
actual_prices.append(prices_list[3] / 2)
actual_prices.append(prices_list[4] / 2)
actual_prices.append(prices_list[5] / 2)
actual_prices.append(prices_list[6] / 2)

print(actual_prices)

[0.995, 1.495, 0.495, 1.995, 0.995, 0.495, 2.995]



Whenever you start doing something like the code above, you should think about using a for loop instead. It'll save you a lot of time and space. You can think of a for loop kind of like doing a copy and paste, but being able to slightly change whatever you're pasting each time you paste it.

### Example 10

In the cell below, create an empty list called `actual_prices`, then using a `for` loop, append to it the actual price that you'll pay for each grocery. I.e., turn the too-long code in the cell above into some nice efficient code using a `for` loop.

In [27]:
actual_prices = []

for price in prices_list:
    actual_prices.append(price / 2) # Divide price by 2 b/c of sale

print(actual_prices)

[0.995, 1.495, 0.495, 1.995, 0.995, 0.495, 2.995]


## Combining control flow statements

Now let's put everything together!

### Example 11

Do what you did in example 8, but only print the grocery and its corresponding price from the `actual_prices` list if the grocery's actual price is less than $1.50

In [28]:
for number in range(number_of_groceries):
    actual_price = prices_list[number] / 2
    
    if actual_price < 1.5:
        print(grocery_list[number], prices_list[number])

eggs 1.99
tofu 2.99
milk 0.99
potatoes 1.99
tortillas 0.99


### Example 12

Now, do you what you did in example 8, but only print the grocery and its corresponding price from the `actual_prices` list if the grocery's actual price is less than $1.50 `or` if that grocery's name is "candy"

In [29]:
for number in range(number_of_groceries):
    actual_price = prices_list[number] / 2
    grocery_name = grocery_list[number]
    
    if actual_price < 1.5 or grocery_name == "candy":
        print(grocery_list[number], prices_list[number])

eggs 1.99
tofu 2.99
milk 0.99
potatoes 1.99
tortillas 0.99
candy 5.99
